### PR TITLE
[x64] make lax_scipy_test.py compatible with strict dtype promotion

### PR DESCRIPTION
**Pull request by jakevdp**
Part of https://github.com/google/jax/pull/10865 and https://github.com/google/jax/pull/10840.

### CODE REVIEW

File: jax/_src/lax/eigh.py [hunk 1]
```diff
@@ -327,6 +327,7 @@ def base_case(B, offset, b, agenda, blocks, eigenvectors):
     # only do the right thing if termination_size == 1.
     H = _mask(H, (b, b), jnp.eye(B, dtype=H.dtype))
     eig_vecs, eig_vals = lax.linalg.eigh(H, sort_eigenvalues=False)
+    eig_vals = eig_vals.astype(eig_vecs.dtype)
```
Comment (hawkinsp) [inline jax/_src/lax/eigh.py:330]: I guess the point of this is that we are embedding the eigenvalues (which are always real) in a complex array. One slight nit is that I might move this type cast after the masking logic.

File: jax/_src/lax/qdwh.py [hunk 2]
```diff
@@ -123,7 +123,7 @@ def _qdwh(x, m, n, is_hermitian, max_iterations, eps):
   if eps is None:
     eps = float(jnp.finfo(x.dtype).eps)
   alpha = (jnp.sqrt(jnp.linalg.norm(x, ord=1)) *
-           jnp.sqrt(jnp.linalg.norm(x, ord=jnp.inf)))
+           jnp.sqrt(jnp.linalg.norm(x, ord=jnp.inf))).astype(x.dtype)
   l = eps
 
   u = x / alpha
```
Comment (hawkinsp) [inline jax/_src/lax/qdwh.py:129]: I'm just reflecting on this. Do we really want to force a cast when doing a `complex / real` division where the `real` type is the same on one of the complex element types?

**Review by hawkinsp — COMMENTED** [review]
I'm not sure how I feel about requiring real -> complex casts!

File: jax/_src/lax/qdwh.py (hunk 2 above)
Comment (jakevdp) [reply jax/_src/lax/qdwh.py:129]: That's a good question – I debated that in https://github.com/google/jax/pull/10824, but landed on the fact that when people say they want `strict` dtype promotion, we should probably avoid any implicit promotion rather than trying to guess which promotions are OK.

Comment (jakevdp) [reply jax/_src/lax/qdwh.py:129]: (I think I even posed this exact question in our x64 chat at some point)

Comment (jakevdp) [reply jax/_src/lax/qdwh.py:129]: I'm going to leave as-is for now; we can always make this less restrictive in the future if we wish.

File: jax/_src/lax/eigh.py [hunk 3]
```diff
@@ -342,6 +343,7 @@ def recursive_case(B, offset, b, agenda, blocks, eigenvectors):
     V = _slice(eigenvectors, (0, offset), (n, b), (N, B))
 
     split_point = jnp.nanmedian(_mask(jnp.diag(jnp.real(H)), (b,), jnp.nan))  # TODO: Improve this?
+    split_point = split_point.astype(H.dtype)
```
[the remark this answers is not in the document]
Comment (jakevdp) [reply jax/_src/lax/eigh.py]: Done, thanks!

File: jax/_src/lax/eigh.py (hunk 1 above)
Comment (jakevdp) [reply jax/_src/lax/eigh.py]: Done

**Review by hawkinsp — APPROVED** [review]
